### PR TITLE
Overhaul Web UI

### PR DESCRIPTION
**Pull request by raghavsethi**
Here's the overhaul of the web UI I worked on over the hackathon and the last couple of weeks.

### Design Goals
- To make it easy to see how a cluster is doing
- To make it easier to find queries in the query list
- To help figure out query performance, stragglers and skew
- To bring the Web UI into 2016

### Non-goals
- Looking decent at window sizes < 1024px
- Supporting browsers that aren't Chrome, Safari or Firefox of a vintage > 3 months ago
- Premature performance optimizations

### New Features
- Cluster heads up display: Now it's easy to see how your cluster is doing, and what sort of load it's under:
  - Running, blocked and queued queries
  - Aggregate data input rates (rows and bytes)
  - Aggregate parallelism (drivers and CPU time)
  - All of these also have sparklines for easy trend analysis
- More powerful query list
  - Finer-grained stats (execution, CPU and elapsed time)
  - Powerful sorting and filtering options
  - Limit number of displayed queries to improve page performance
  - Regex search on query ID, source, user and query text
- Completely re-written query detail page
  - Query resource utilization statistics, and sparklines for parallelism, input data and memory
  - All new stage-level statistics
    - See timings, memory and task info
    - Find stragglers and outliers with bar charts of CPU time and buffers for all tasks
  - And best of all - everything on this page is live - you can see stats change as a query progresses (can also be paused)
- All UI code has been either re-written or cleaned up dramatically. Everything is now in React.

### Regressions
- The new table framework doesn't support sticky headers out of the box (will add it back eventually)
- Sorting the tasks table by multiple columns sequentially does not maintain previous sort orders
- Removed the stage view in the query detail page (it seemed to be universally despised)
- The density of information on the cluster overview page seems like a potential regression to me, but the demand for more information won the battle.
- There are about 5x more bells and whistles than before - which means that I'm going to be fixing edge cases for weeks.

### Feedback Required
cc @martint @dain @haozhun @electrum @nileema 
- I've made dozens of copy and design decisions, please let me know if (a) you feel strongly about something or (b) it's easy to change.
- I need to figure out how to roll this out in a way that we're not stuck with a broken UI in between releases. My current idea is to have a banner at the top of the old UI with a link to the new one first. Then I would change it to the new UI and put a link to the old one. Then I would remove the old one completely. Is this too complex? Can I just bite the bullet and replace the old UI?

As an aside, I've tried to avoid adding dependencies, especially into the build process. This means that I rejected all libraries that really wanted me to have Webpack or Browserify, or ones that the only way to get a piece of JS out of was to install a CoffeeScript-JSX Compiler (!!). This also means that we're using JSX, which is not great for performance, but I'm going to punt on compiling stuff until we know there's a perf problem.

### CODE REVIEW

**Comment by electrum** [comment]
Can we keep the color logo?

**Comment by electrum** [comment]
Please split this up to add the new vendor resources as a separate commit. This will make it easier to review and look at later.

**Comment by raghavsethi** [comment]
Split into two commits that will be squashed before merge.

- Changed to color logo
- Added histograms
- Made query page sparklines show instantaneous rather than average stats

**Comment by cberner** [comment]
Can you split up the first commit into 1) make copy of UI under /old/, 2) add new libraries? That will make it easier to review, as right now there's just a 81k line commit and it's not clear what's happening :) Also the commits need better titles and descriptions

**Comment by raghavsethi** [comment]
Restructured the PR to make it easier to review, also made a whole bunch of perf improvements.

File: presto-main/src/main/resources/webapp/vendor/bootstrap/config.json [hunk 1]
```diff
@@ -0,0 +1,435 @@
+{
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/vendor/bootstrap/config.json]: List in the commit message what libraries are being added.

Comment (cberner) [inline presto-main/src/main/resources/webapp/vendor/bootstrap/config.json]: Also, are you sure all the necessary license files are included? It seems like the license for Solarized, at least, is missing: https://github.com/altercation/solarized/blob/master/LICENSE

**Comment by raghavsethi** [comment]
Listed libraries in the commit, and added license files for all libraries that were missing them.

File: presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java [hunk 2]
```diff
@@ -192,6 +204,36 @@ public Duration getElapsedTime()
...
+    }
+
+    @JsonProperty
+    public Duration getCpuTime()
+    {
+        return cpuTime;
+    }
+
+    @JsonProperty
+    public long getExecutionTimeMillis()
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java]: Why do we have both `getExecutionTime` and `getExecutionTimeMillis`?

Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java]: To maintain compatibility for the old UI. Should I change the old one instead?

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java]: I'm a bit confused. Why do we need both for backwards compatibility? Both of these methods are new

Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java]: Sorry I confused this with another method in a an earlier version of the PR. I'll get rid of the millis code.

**Comment by ArturGajowy** [comment]
I propose using http://www.webjars.org/ for managing frontend dependencies

File: presto-main/src/main/resources/webapp/assets/cluster.js [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+        </div>);
+    }
+});
+
+ReactDOM.render(
+        <ClusterHUD />,
+        document.getElementById('cluster-hud')
+);
+
+var Query = React.createClass({
```
Comment (ArturGajowy) [inline presto-main/src/main/resources/webapp/assets/cluster.js]: Maybe let's split this file so that each react component has its own file?

Comment (ArturGajowy) [inline presto-main/src/main/resources/webapp/assets/cluster.js]: This also concerns assets/query.js

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                }
+            }
+
+            if (!query.getState().equals(QueryState.FAILED) && !query.getState().equals(QueryState.FINISHED)) {
+                double totalExecutionTimeSeconds = query.getQueryStats().getElapsedTime().toMillis() / 1000.0;
+                if (totalExecutionTimeSeconds != 0) {
+                    byteInputRate += query.getQueryStats().getProcessedInputDataSize().toBytes() / totalExecutionTimeSeconds;
+                    rowInputRate += query.getQueryStats().getProcessedInputPositions() / totalExecutionTimeSeconds;
+                    cpuTimeRate += (query.getQueryStats().getTotalCpuTime().toMillis() / 1000.0) / totalExecutionTimeSeconds;
+                    memoryReservation += query.getQueryStats().getTotalMemoryReservation().toBytes();
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: Instead of total "user" memory. It might be more useful to display free distributed memory in the general pool & system pool, along with min local memory on a node for general & system

File: presto-main/src/main/java/com/facebook/presto/server/QueryResource.java [hunk 5]
```diff
@@ -68,16 +69,16 @@ public QueryResource(QueryManager queryManager, AccessControl accessControl, Ses
     }
 
     @GET
-    public List<BasicQueryInfo> getAllQueryInfo()
+    public Map<String, BasicQueryInfo> getAllQueryInfo()
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/QueryResource.java]: Do we need this change? It seems like the client can just index the infos on the client side, if they want a map

File: presto-main/src/main/resources/webapp/assets/cluster.js (hunk 3 above)
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/cluster.js]: Ya, I agree. This file is pretty long

File: presto-main/src/main/resources/webapp/assets/query.js [hunk 6]
```diff
@@ -0,0 +1,1087 @@
...
+    Td = Reactable.Td;
+
+var TaskList = React.createClass({
+    getTasks: function (stage) {
+        return [].concat.apply(stage.tasks, stage.subStages.map(this.getTasks));
+    },
+    render: function() {
+        var tasks = this.getTasks(this.props.outputStage);
+        var renderedTasks = tasks.map(function (task) {
+            var t = task;
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/query.js]: avoid single letter variables

File: presto-main/src/main/resources/webapp/assets/query.js [hunk 7]
```diff
@@ -0,0 +1,1087 @@
...
+            var elapsedTime = parseDuration(t.stats.elapsedTime);
+            if (elapsedTime == 0) {
+                elapsedTime = Date.now() - Date.parse(t.stats.createTime);
+            }
+            return (
+                    <Tr key={t.taskStatus.taskId}>
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/query.js]: Seems like all the other tags are lower case, so I would stick with that for "Tr" and "Td" too

File: presto-main/src/main/resources/webapp/assets/query.js [hunk 8]
```diff
@@ -0,0 +1,1087 @@
...
+                        </Td>
+                        <Td column="bufferedBytes" value={t.outputBuffers.totalBufferedBytes}>
+                            {formatDataSizeBytes(t.outputBuffers.totalBufferedBytes)}
+                        </Td>
+                    </Tr>
+            );
+        }.bind(this));
+
+        return (
+            <Table id="tasks" className="table table-striped sortable" sortable={true} defaultSort={{column: 'id', direction: 'asc'}}>
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/query.js]: lower case

File: presto-main/src/main/resources/webapp/assets/query.js [hunk 9]
```diff
@@ -0,0 +1,1087 @@
...
+                <Thead>
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/query.js]: lower case for all of these

File: presto-main/src/main/resources/webapp/assets/query.js [hunk 10]
```diff
@@ -0,0 +1,1087 @@
...
+        var tooltipValueLookups = {'offset' : {}};
+        for (var i = 0; i < histogramData.length; i++) {
+            tooltipValueLookups['offset'][i] = numberFormatter(dataMin + (i * bucketSize)) + "-" + numberFormatter(dataMin + ((i + 1) * bucketSize));
+        }
+
+        var stageHistogramProperties = $.extend({}, HISTOGRAM_PROPERTIES,  {barWidth: (HISTOGRAM_WIDTH / histogramData.length), tooltipValueLookups: tooltipValueLookups});
+        $(histogramId).sparkline(histogramData, stageHistogramProperties);
+    },
+    componentDidUpdate: function() {
+        var s = this.props.stage;
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/query.js]: avoid single letter variable names

File: presto-main/src/main/resources/webapp/assets/query.js (hunk 9 above)
Comment (raghavsethi) [inline presto-main/src/main/resources/webapp/assets/query.js]: This casing comes from the Reactable library

File: presto-main/src/main/resources/webapp/assets/query.js [hunk 11]
```diff
@@ -0,0 +1,1087 @@
...
+    render: function() {
+        var s = this.props.stage;
+
+        var totalBufferedBytes = s.tasks.map(function(task) {
+            return task.outputBuffers.totalBufferedBytes;
+        }).reduce(function(previousValue, currentValue) {
+            return previousValue + currentValue;
+        });
+
+        return (
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/query.js]: This block of HTML is pretty long. Maybe there's a way to split it up a bit more?

File: presto-main/src/main/resources/webapp/assets/utils.js [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+var EWMA_ALPHA = 0.2;
+
+function addToHistory (value, valuesArray) {
+    if (valuesArray.length == 0) {
+        return valuesArray.concat([value]);;
+    }
+    return valuesArray.concat([value]).slice(Math.max(valuesArray.length - MAX_HISTORY, 0));
+}
+
+function addEwmaToHistory (value, valuesArray) {
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/utils.js]: What's a Ewma? Is there a better name for this?

**Comment by cberner** [comment]
Seems like there's something wrong with the time rendering:
screen shot 2016-05-24 at 11 04 16 am

**Comment by cberner** [comment]
I found it hard to tell which buttons were selected:
screen shot 2016-05-24 at 11 06 42 am

Maybe check with a few others to see if it's just me?

**Comment by cberner** [comment]
Cumulative memory looks wrong:
screen shot 2016-05-24 at 11 09 26 am

How can it be non-zero when the peak is zero?

**Comment by cberner** [comment]
I think displaying the current memory usage for each query on the main query list would be useful, and being able to sort by that would be great. A somewhat common case is wanting to find the biggest query by memory so that you can kill it manually to free up the cluster

**Comment by cberner** [comment]
Back to you for some comments

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java (hunk 4 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: Or maybe also a % of free memory

File: presto-main/src/main/resources/webapp/assets/cluster.js [hunk 13]
```diff
@@ -0,0 +1,740 @@
...
+            $('#cpu-time-rate-sparkline').sparkline(this.state.cpuTimeRate, $.extend({}, SPARKLINE_PROPERTIES, {numberFormatter: precisionRound}));
+
+            this.setState({
+                lastRender: renderTimestamp
+            });
+        }
+
+        $('[data-toggle="tooltip"]').tooltip();
+    },
+    render: function() {
```
Comment (sopel39) [inline presto-main/src/main/resources/webapp/assets/cluster.js]: Is it possible to separate renedered templates from classes in React? If so, please move templates to separate files. It would increase code readability.

File: presto-main/src/main/resources/webapp/assets/cluster.js [hunk 14]
```diff
@@ -0,0 +1,740 @@
...
+var Query = React.createClass({
+    render: function()
+    {
+        var query = this.props.query;
+        var summary = query.state;
+        var progress = 0;
+
+        var completedDrivers = query.completedDrivers;
+        var runningDrivers = query.runningDrivers;
+        var queuedDrivers = query.queuedDrivers;
```
Comment (sopel39) [inline presto-main/src/main/resources/webapp/assets/cluster.js]: I would guess it should be possible to precompute all variables required for rendering and pass them to template engine. Is it possible in React?

File: presto-main/src/main/resources/webapp/old/units.js [hunk 15]
```diff
@@ -16,8 +16,8 @@ function formatDuration(millis) {
     return precisionRound(millis) + " " + unit;
 }
 
-function formatDataSize(bytes) {
-    var unit = "B";
+function formatDataSize(bytes, defaultUnit = "B") {
```
Comment (sopel39) [inline presto-main/src/main/resources/webapp/old/units.js]: How is this used? What if `defaultUnit != "B"` and `bytes < 1024`. What is example output?

File: presto-main/src/main/resources/webapp/assets/cluster.js (hunk 3 above)
Comment (raghavsethi) [inline presto-main/src/main/resources/webapp/assets/cluster.js]: Breaking this up into multiple files that have dependencies on each other means that we will have to either do some sort of ugly hack to pass globals around or use something like Webpack. I think having files be longer than we like is the lesser of the two evils.

File: presto-main/src/main/resources/webapp/assets/query.js (hunk 11 above)
Comment (raghavsethi) [inline presto-main/src/main/resources/webapp/assets/query.js]: I believe that splitting this up makes it harder to maintain. All the code is essentially just layout, and putting stuff in separate methods would make it harder to reason about the resultant DOM.

File: presto-main/src/main/resources/webapp/assets/cluster.js (hunk 13 above)
Comment (raghavsethi) [inline presto-main/src/main/resources/webapp/assets/cluster.js]: React doesn't really do templates. AFAIK this is idiomatic React.

File: presto-main/src/main/resources/webapp/assets/cluster.js (hunk 14 above)
Comment (raghavsethi) [inline presto-main/src/main/resources/webapp/assets/cluster.js]: I've either precomputed everything that is complex or factored it out into separate functions.

File: presto-main/src/main/resources/webapp/old/units.js (hunk 15 above)
Comment (raghavsethi) [inline presto-main/src/main/resources/webapp/old/units.js]: This change shouldn't have been committed at all. Thanks for pointing it out :)

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java (hunk 4 above)
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: This seems like it needs Java code changes and more thought about how to get that information. I'm going to create a separate issue to track that feature request.

**Comment by raghavsethi** [comment]
@cberner back to you. It's also running in the test cluster so you can play with it.

File: presto-main/src/main/resources/webapp/assets/utils.js [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+            return QUERY_STATE_COLOR_MAP.QUEUED;
+    }
+};
+
+// Sparkline-related functions
+// ===========================
+
+// display at most 5 minutes worth of data on the sparklines
+var MAX_HISTORY = 60 * 5;
+ // alpha param of exponentially weighted modving average. picked arbitrarily - lower values means more smoothness
```
Comment (sopel39) [inline presto-main/src/main/resources/webapp/assets/utils.js]: modving -> moving

File: presto-main/src/main/resources/webapp/assets/utils.js [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+    if (pos != -1) {
+        return id.substring(pos + 1);
+    }
+
+    return id
+}
+
+function taskIdToSortableString(id) {
+    var taskIdArr = removeQueryId(id).split(".");
+    return taskIdArr.map( function(numStr) { return new Array(5 - numStr.length).join("0") + numStr }).join("|");
```
Comment (sopel39) [inline presto-main/src/main/resources/webapp/assets/utils.js]: why 5? This formula seems a bit complicated.
Please add an example `id` so reader can understand this transformation

File: presto-main/src/main/resources/webapp/assets/utils.js [hunk 18]
```diff
@@ -0,0 +1,387 @@
...
+        unit = "T";
+    }
+    if (count > 1000) {
+        count /= 1000;
+        unit = "Q";
+    }
+    return precisionRound(count) + unit;
+}
+
+function formatDataSizeBytes(size) {
```
Comment (sopel39) [inline presto-main/src/main/resources/webapp/assets/utils.js]: Is it necessary to have almost duplicate function of `formatDataSize`?

**Comment by sopel39** [comment]
Possible UI bugs when running `explain analyze select * from tpch.sf1.orders,tpch.sf10.orders;` query. Browser is Chrome

When clicking `Live Plan` on query page:
redirected to: `http://localhost:8080/ui/plan?20160602_111303_00011_cn56s`
error:
```
http://localhost:8080/vendor/dagre-d3/dagre-d3-0.4.14.js Failed to load resource: the server responded with a status of 404 (Not Found)
http://localhost:8080/units.js Failed to load resource: the server responded with a status of 404 (Not Found)
plan?20160602_111303_00011_cn56s:212 Uncaught ReferenceError: dagreD3 is not defined(anonymous function) @ plan?20160602_111303_00011_cn56s:212
http://localhost:8080/favicon.ico Failed to load resource: the server responded with a status of 404 (Not Found)
```

When clicking `Split Timeline`
redirected to: `http://localhost:8080/timeline.html?20160602_111303_00011_cn56s`
error:
```
timeline.html?20160602_111303_00011_cn56s:124 Uncaught TypeError: Cannot read property 'substring' of undefined
```
I think ` task.taskId` is `undefined`

**Comment by sopel39** [comment]
zrzut ekranu 2016-06-02 o 13 31 48

I noticed that there is no UI way of going back from deeper page (e.g. query page) to more top level page (e.g. main page). The only way to go to top level page is by using browser `back` button.

I would suggest to add breadcrumb navigation bar at the top of each page. It would basically show clickable link to the higher level pages, e.g:
`Queries -> Query 123 -> Slice Timeline`.
See http://marketingland.com/breadcrumb-links-good-user-experience-yes-97848

File: presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java [hunk 19]
```diff
@@ -70,8 +76,13 @@ public BasicQueryInfo(
             @JsonProperty("self") URI self,
             @JsonProperty("query") String query,
             @JsonProperty("elapsedTime") Duration elapsedTime,
+            @JsonProperty("executionTime") Duration executionTime,
+            @JsonProperty("cpuTime") Duration cpuTime,
             @JsonProperty("endTime") DateTime endTime,
             @JsonProperty("createTime") DateTime createTime,
+            @JsonProperty("currentMemory") DataSize currentMemory,
+            @JsonProperty("currentMemory") DataSize peakMemory,
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java]: typo

File: presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java [hunk 20]
```diff
@@ -186,9 +207,21 @@ public String getQuery()
...
+    }
+
+    @JsonProperty
+    public long getCpuTimeMillis()
+    {
+        return cpuTime.toMillis();
+    }
+
+    @JsonProperty
+    public long getElapsedTimeMillis()
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java]: This doesn't look like it's going to round trip anymore. On the other hand, I'm not sure we ever deserialize these. Either way though, either remove the JSON constructor or fix these so that it round-trips

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+            if (query.getState().equals(QueryState.RUNNING)) {
+                if (query.getQueryStats().isFullyBlocked()) {
+                    blockedQueries++;
+                }
+                else {
+                    runningQueries++;
+                }
+            }
+
+            if (!query.getState().equals(QueryState.FAILED) && !query.getState().equals(QueryState.FINISHED)) {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: I think `!query.getState().isDone()` would be better

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java (hunk 4 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: Should all 5 of these be inside the `if` block? It seems like only the first 3 should be

File: presto-main/src/main/resources/webapp/assets/presto.css [hunk 22]
```diff
@@ -0,0 +1,551 @@
...
+    width: 100px;
+    padding-right: 5px;
+}
+
+.expanded-chart {
+    display: inline-table;
+}
+
+/**
+The MIT License (MIT)
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/presto.css]: Why is this here? If some of the code is copied from somewhere else, I would move it to a separate file

File: presto-main/src/main/resources/webapp/query.html [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+                        <td>
+                            <span class="navbar-brand">Query Details</span>
+                        </td>
+                    </tr>
+                </table>
+            </div>
+            <div id="navbar" class="navbar-collapse collapse">
+                <ul class="nav navbar-nav navbar-right">
+                    <li>
+                            <span class="navbar-cluster-info">
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/query.html]: Spacing looks off here

File: presto-main/src/main/resources/webapp/query.html [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+                                Version <br/>
+                                <span class="text" id="version-number">...</span>
+                            </span>
+                    </li>
+                    <li>
+                            <span class="navbar-cluster-info">
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/query.html]: and here

File: presto-main/src/main/resources/webapp/index.html [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+                        <td>
+                            <span class="navbar-brand">Cluster Overview</span>
+                        </td>
+                    </tr>
+                </table>
+            </div>
+            <div id="navbar" class="navbar-collapse collapse">
+                <ul class="nav navbar-nav navbar-right">
+                    <li>
+                            <span class="navbar-cluster-info">
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/index.html]: Spacing

File: presto-main/src/main/resources/webapp/index.html [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+                                Version <br/>
+                                <span class="text" id="version-number">...</span>
+                            </span>
+                    </li>
+                    <li>
+                            <span class="navbar-cluster-info">
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/index.html]: and here

**Comment by cberner** [comment]
Back to you for comments

**Comment by raghavsethi** [comment]
@sopel39 thanks for your comments. I'm going to leave breadcrumb navigation as something we can explore later, you can click the Presto logo to go back to the cluster overview page for now. #5129 is already tracking the split timeline bug.

@cberner back to you.

File: presto-main/src/main/resources/com/facebook/presto/server/plan.html [hunk 27]
```diff
@@ -4,7 +4,7 @@
 
     <script src="/vendor/d3/d3-3.3.4.js"></script>
     <script src="/vendor/dagre-d3/dagre-d3-0.4.14.js"></script>
-    <script src="/units.js"></script>
+    <script src="/assets/utils.js"></script>
```
Comment (cberner) [inline presto-main/src/main/resources/com/facebook/presto/server/plan.html]: Seems like this change should be in a different commit

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java [hunk 28]
```diff
@@ -0,0 +1,179 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.inject.Inject;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/")
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: Make this `/v1/cluster` in the style of `QueryResource`

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java (hunk 4 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: I think you missed this comment

File: presto-main/src/main/resources/webapp/assets/cluster-hud.js [hunk 29]
```diff
@@ -0,0 +1,259 @@
...
+        .error(function() {
+            this.resetTimer();
+        }.bind(this));
+    },
+    componentDidMount: function() {
+        this.refreshLoop();
+    },
+    componentDidUpdate: function() {
+        // prevent multiple calls to setState within the refresh interval from re-rendering sparklines/charts
+        if (this.state.lastRender == null || (Date.now() - this.state.lastRender) >= 1000) {
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/cluster-hud.js]: This `lastRender` looks hacky. Why do we need it?

File: presto-main/src/main/resources/webapp/assets/loader.css [hunk 30]
```diff
@@ -0,0 +1,64 @@
+/**
+The MIT License (MIT)
+
+Copyright (c) 2014 Luke Haas
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/loader.css]: Should this file be in the vendor directory? I assume it came from somewhere else

File: presto-main/src/main/resources/webapp/assets/query-list.js [hunk 31]
```diff
@@ -0,0 +1,509 @@
...
+                    }
+                }
+                else {
+                    summary = (progress == 0 ? summary : summary + " (" + progress + "%" + ")");
+                }
+                break;
+            case "FINISHED":
+                progress = 100;
+                runningDrivers = 0;
+                queuedDrivers = 0;
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/query-list.js]: Add `break;`

File: presto-main/src/main/resources/webapp/assets/query-list.js [hunk 32]
```diff
@@ -0,0 +1,509 @@
...
+                    <span className="glyphicon glyphicon-ok" style={GLYPHICON_HIGHLIGHT}></span>&nbsp;&nbsp;
+                    { completedDrivers }
+                </span>
+                <span className="tinystat" data-toggle="tooltip" data-placement="top" title="Running splits">
+                    <span className="glyphicon glyphicon-play" style={GLYPHICON_HIGHLIGHT}></span>&nbsp;&nbsp;
+                    { runningDrivers }
+                </span>
+                <span className="tinystat" data-toggle="tooltip" data-placement="top" title="Queued splits">
+                    <span className="glyphicon glyphicon-pause" style={GLYPHICON_HIGHLIGHT}></span>&nbsp;&nbsp;
+                    { queuedDrivers }
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/query-list.js]: You're inconsistent about having or not having spaces before/after the braces. Make them consistent

File: presto-main/src/main/resources/webapp/assets/utils.js [hunk 33]
```diff
@@ -0,0 +1,367 @@
...
+    }
+    return valuesArray.concat([value]).slice(Math.max(valuesArray.length - MAX_HISTORY, 0));
+}
+
+function addExponentiallyWeightedToHistory (value, valuesArray) {
+    if (valuesArray.length == 0) {
+        return valuesArray.concat([value]);;
+    }
+
+    var ewma = (value * EWMA_ALPHA) + (valuesArray[valuesArray.length - 1] * (1 - EWMA_ALPHA));
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/utils.js]: avoid abbreviations. Maybe `movingAverage`?

File: presto-main/src/main/resources/webapp/assets/utils.js [hunk 34]
```diff
@@ -0,0 +1,367 @@
...
+    if (fullyBlocked && state == "RUNNING") {
+        return "BLOCKED";
+    }
+    else {
+        return state;
+    }
+}
+
+function getHostname(hostname) {
+    var s = new URL(hostname).hostname;
```
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/utils.js]: avoid single letter variables

File: presto-main/src/main/resources/webapp/assets/utils.js (hunk 17 above)
Comment (cberner) [inline presto-main/src/main/resources/webapp/assets/utils.js]: Can React take a custom comparator instead? I think that would make it easier to understand

**Comment by cberner** [comment]
Had a few more comments

**Comment by raghavsethi** [comment]
Back to you

**Comment by cberner** [comment]
Have @dain, @martint, or @electrum look at the REST API changes. The rest looks good

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java [hunk 35]
```diff
@@ -239,6 +241,16 @@ public Duration getQueuedTime()
     }
 
     @JsonProperty
+    public Duration getExecutionTime()
+    {
+        if (queuedTime == null) {
+            // counter-intuitively, this means that the query is still queued
+            return new Duration(0, TimeUnit.NANOSECONDS);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStats.java]: Static import `NANOSECONDS`

File: presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java [hunk 36]
```diff
@@ -227,6 +257,24 @@ public DateTime getCreateTime()
...
     }
 
+    @JsonProperty
+    public double getCumulativeMemory()
+    {
+        return cumulativeMemory;
+    }
+
+    @JsonProperty
+    public DataSize getCurrentMemory()
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/BasicQueryInfo.java]: Make these bytes

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java [hunk 37]
```diff
@@ -0,0 +1,178 @@
...
+@Path("/v1/cluster")
+public class ClusterStatsResource
+{
+    private final InternalNodeManager nodeManager;
+    private final QueryManager queryManager;
+
+    @Inject
+    public ClusterStatsResource(InternalNodeManager nodeManager, QueryManager queryManager)
+    {
+        this.nodeManager = nodeManager;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: Check for null

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java [hunk 38]
```diff
@@ -0,0 +1,178 @@
...
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/v1/cluster")
+public class ClusterStatsResource
+{
+    private final InternalNodeManager nodeManager;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: These can all be `NodeManager`

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java [hunk 39]
```diff
@@ -0,0 +1,178 @@
...
+        long activeNodes = nodeManager.getNodes(NodeState.ACTIVE).size();
+        long runningDrivers = 0;
+        double memoryReservation = 0;
+
+        double rowInputRate = 0;
+        double byteInputRate = 0;
+        double cpuTimeRate = 0;
+
+        for (QueryInfo query : queryManager.getAllQueryInfo()) {
+            if (query.getState().equals(QueryState.QUEUED)) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: Enums can use `==`

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java [hunk 40]
```diff
@@ -0,0 +1,178 @@
...
+                queuedQueries++;
+            }
+            if (query.getState().equals(QueryState.RUNNING)) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: This can be `else`

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java [hunk 41]
```diff
@@ -0,0 +1,178 @@
...
+                if (query.getQueryStats().isFullyBlocked()) {
+                    blockedQueries++;
+                }
+                else {
+                    runningQueries++;
+                }
+            }
+
+            if (!query.getState().isDone()) {
+                double totalExecutionTimeSeconds = query.getQueryStats().getElapsedTime().toMillis() / 1000.0;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: I think this should be `.getElapsedTime().getValue(SECONDS)`

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java [hunk 42]
```diff
@@ -0,0 +1,178 @@
...
+                if (totalExecutionTimeSeconds != 0) {
+                    byteInputRate += query.getQueryStats().getProcessedInputDataSize().toBytes() / totalExecutionTimeSeconds;
+                    rowInputRate += query.getQueryStats().getProcessedInputPositions() / totalExecutionTimeSeconds;
+                    cpuTimeRate += (query.getQueryStats().getTotalCpuTime().toMillis() / 1000.0) / totalExecutionTimeSeconds;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: Same

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java [hunk 43]
```diff
@@ -0,0 +1,178 @@
...
+package com.facebook.presto.server;
+
+import com.facebook.presto.execution.QueryInfo;
+import com.facebook.presto.execution.QueryManager;
+import com.facebook.presto.execution.QueryState;
+import com.facebook.presto.metadata.InternalNodeManager;
+import com.facebook.presto.spi.NodeState;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.inject.Inject;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java]: Use `javax.inject` instead
